Code/Framework/AzCore/AzCore/Task/TaskExecutor.cpp

```diff
@@ -59,30 +59,33 @@ namespace AZ
 
         uint32_t CompiledTaskGraph::Release()
         {
+            // Release is run from many threads, and another thread can azdestroy(this) as soon as the remaining count is decremented.
+            // READ ALL NECESSARY DATA BEFORE DECREMENTING THE REMAINING COUNT!
+            bool isRetained = m_parent != nullptr;
+            TaskGraph* parent = m_parent;
```

Comment:
When does m_parent get set to nullptr?

Reply:
My concern is that there can be a scenario where this Release function is called with a ref count of 2, by the two remaining references and the m_parent is non-nullptr.

Thread A(first ref):
* Sets `isRetained` = true
* stores off a local reference to m_parent into `parent`
* stores off local reference to m_waitEvent into `waitEvent`
* Atomically decrements m_remaining to 1

Now Thread B runs(second ref):
* Sets `isRetained` = true
* stores off a local reference to m_parent into `parent`
* stores off local reference to m_waitEvent into `waitEvent`
* Atomically decrements m_remaining to 0
* Goes into the `if (isRetained)` block, but does nothing because remaining is 0.
* returns 0

Thread A then continues:
* Goes into the `if (isRetained)` block and succeeds the `if (remaining == 1)` check
* Sets parent->m_submitted to false
* Signals the `waitEvent` if it is non-nullptr
* returns 1

In that case the CompiledTaskGraph is never deleted.

Reply:
Thanks for the clarification.